### PR TITLE
Storage/sc  search api

### PR DESCRIPTION
**Pull request by spencerc99**
## Summary
* added search api handler and required models
* added method in SQLCommandController to execute arbitrary SQL statement

## Testing Plan
* unit tests
* tested in postman
image

## Affected Services
storage

## Outstanding Potential Issues/Problems
* can't pass in -1 or whatever we said to pass in for start and end dates to only get one
* no error checking for date stuff

To fix, PROBABLY should change design of endpoint to have explicit keys for start and end dates so if only one is present then doesn't need to do in between.

## Reviewers
@Comp413-F18/storage

### CODE REVIEW

File: Storage/src/main/java/storage/SearchHandler.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        if (body.getFilter().getKeywords().isPresent()) {
+            String keywordsFilter = List.of(body.getFilter().getKeywords().get()).stream()
+                .map((keyword) -> String.format("description LIKE '%%s%'", keyword))
+                .reduce((str1, str2) -> String.format("%s OR %s", str1, str2)).get();
+            selectStatement.append(keywordsFilter);
+        }
+
+        if (body.getFilter().getCoverageRange().isPresent()) {
+            Timestamp[] coverageRange = body.getFilter().getCoverageRange().get();
+            // add check for strictly before or after some range, for now assumes strictly between range
```
Comment (willkoh) [inline Storage/src/main/java/storage/SearchHandler.java]: I think we can do this and tell UI/UX to send the first UNIX timestamp for start range if strictly before or today's date for end range if strictly after.

File: Storage/src/main/java/storage/controllers/DatabaseCRUDController.java [hunk 2]
```diff
@@ -1,11 +1,23 @@
...
 
+import org.json.JSONArray;
 import org.json.JSONException;
+import org.springframework.http.ResponseEntity;
 import org.springframework.web.bind.annotation.*;
 import storage.models.MediaMention;
 import storage.models.Response;
 
 import java.sql.SQLException;
+import com.sun.tools.javac.util.List;
```
Comment (willkoh) [inline Storage/src/main/java/storage/controllers/DatabaseCRUDController.java]: Again, why this `List` over `java.util.List`?

File: Storage/src/main/java/storage/controllers/SQLCommandController.java [hunk 3]
```diff
@@ -143,6 +143,13 @@ public JSONArray executeSelectAllStatement() throws SQLException, JSONException
         return result;
     }
 
+    public JSONArray executeCustomSelectStatement(String selectStatement) throws SQLException, JSONException {
+        ResultSet rset = stmt.executeQuery(selectStatement);
+        JSONArray result = ResultSetConverter.convert(rset);
+        return result;
+
```
Comment (willkoh) [inline Storage/src/main/java/storage/controllers/SQLCommandController.java]: Can we remove this empty line?

File: Storage/src/main/java/storage/models/MediaMentionUniqueAttributes.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
+package storage.models;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+
+import java.util.HashMap;
+import java.util.Optional;
```
Comment (willkoh) [inline Storage/src/main/java/storage/models/MediaMentionUniqueAttributes.java]: `Optionals` remind me of 215 :sad_dabs:

File: Storage/src/main/java/storage/models/MediaMentionUniqueAttributes.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    private Optional<Boolean> datelineDateUnique;
+    private Optional<Boolean> categoryUnique;
+    @JsonIgnore
+    private HashMap<String, Optional> allValues;
+
+    public Optional<Boolean> getUrlUnique() {
+        return urlUnique;
+    }
+
+    public void setUrlUnique(Optional<Boolean> urlUnique) {
```
Comment (willkoh) [inline Storage/src/main/java/storage/models/MediaMentionUniqueAttributes.java]: Might be more useful to take `Boolean` as input and construct an `Optional<Boolean>` from that bool.

File: Storage/src/main/java/storage/models/MediaMentionUniqueAttributes.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    private Optional<Boolean> mediaOutletNameUnique;
+    private Optional<Boolean> mediaTypeUnique;
+    private Optional<Boolean> scopeUnique;
+    private Optional<Boolean> localeUnique;
+    private Optional<Boolean> languageUnique;
+    private Optional<Boolean> coverageDateUnique;
+    private Optional<Boolean> datelineDateUnique;
+    private Optional<Boolean> categoryUnique;
+    @JsonIgnore
+    private HashMap<String, Optional> allValues;
```
Comment (willkoh) [inline Storage/src/main/java/storage/models/MediaMentionUniqueAttributes.java]: Could we keep all of this in `allValues` and simply access from that map?

File: Storage/src/main/java/storage/models/MediaMentionUniqueAttributes.java (hunk 5 above)
Comment (willkoh) [reply Storage/src/main/java/storage/models/MediaMentionUniqueAttributes.java]: Or is it a Spring thing to do it this way to ensure correct access to those fields?

File: Storage/src/main/java/storage/SearchHandler.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
+package storage;
+
+import com.sun.tools.javac.util.List;
```
[the remark this answers is not in the document]
Comment (willkoh) [reply Storage/src/main/java/storage/SearchHandler.java]: That's a specific oracle package, so I feel like it was probably suggested by your IDE.

File: Storage/src/main/java/storage/SearchHandler.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        if (body.getFilter().getAuthors().isPresent()) {
+            String authorFilter = List.of(body.getFilter().getAuthors().get()).stream()
+                .map((author) -> String.format("'%s' IN authors", author))
+                .reduce((str1, str2) -> String.format("%s OR %s", str1, str2)).get();
+            selectStatement.append(" AND ").append(authorFilter);
+        }
+
+        if (body.getFilter().getKeywords().isPresent()) {
+            String keywordsFilter = List.of(body.getFilter().getKeywords().get()).stream()
+                .map((keyword) -> String.format("description LIKE '%%s%'", keyword))
```
Comment (willkoh) [inline Storage/src/main/java/storage/SearchHandler.java]: https://stackoverflow.com/questions/759580/how-to-implement-a-keyword-search-in-mysql"

^ hmmm wondering if we use InnoDB Engine or MyISAM engine.

Comment (spencerc99) [reply Storage/src/main/java/storage/SearchHandler.java]: idk tbd later ill add to trello card

File: Storage/src/main/java/storage/models/MediaMentionUniqueAttributes.java (hunk 6 above)
Comment (spencerc99) [reply Storage/src/main/java/storage/models/MediaMentionUniqueAttributes.java]: what do u mean all of this

File: Storage/src/main/java/storage/models/MediaMentionUniqueAttributes.java (hunk 4 above)
Comment (spencerc99) [reply Storage/src/main/java/storage/models/MediaMentionUniqueAttributes.java]: lmao

File: Storage/src/main/java/storage/controllers/DatabaseCRUDController.java (hunk 2 above)
Comment (spencerc99) [reply Storage/src/main/java/storage/controllers/DatabaseCRUDController.java]: because this is the list u need to do streams with

Comment (spencerc99) [reply Storage/src/main/java/storage/controllers/DatabaseCRUDController.java]: idk why but this is the one with List.of if u want i can refactor later to java.util.list i feel like i used this before but maybe not

seems like i could do this
`new ArrayList<>(Arrays.asList(array))`

File: Storage/src/main/java/storage/SearchHandler.java (hunk 1 above)
Comment (spencerc99) [reply Storage/src/main/java/storage/SearchHandler.java]: differnet suggestion in pr description

File: Storage/src/main/java/storage/SearchHandler.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        selectStatement.append(selection).append(" FROM MediaMentions");
+        String filter = body.getFilter().getAllSingleValues().entrySet().stream()
+            .filter((e) -> e.getValue().isPresent())
+            .map((e) -> String.format("%s = '%s'", e.getKey(),  e.getValue().get().toString()))
+            .reduce((str1, str2) -> String.format("%s AND %s", str1, str2)).get();
+        selectStatement.append(" WHERE ").append(filter);
+        if (body.getFilter().getAuthors().isPresent()) {
+            String authorFilter = List.of(body.getFilter().getAuthors().get()).stream()
+                .map((author) -> String.format("'%s' IN authors", author))
+                .reduce((str1, str2) -> String.format("%s OR %s", str1, str2)).get();
```
[the remark this answers is not in the document]
Comment (spencerc99) [reply Storage/src/main/java/storage/SearchHandler.java]: ya thats what we decided ORs on filters with arrays i thought

File: Storage/src/main/java/storage/SearchHandler.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+import com.sun.tools.javac.util.List;
+import storage.models.SearchPayload;
+
+import java.sql.Driver;
+import java.sql.Timestamp;
+
+/**
+ * Created by spencerchang on 10/5/18.
+ */
+public class SearchHandler {
```
[the remark this answers is not in the document]
Comment (spencerc99) [reply Storage/src/main/java/storage/SearchHandler.java]: moves this

File: Storage/src/main/java/storage/models/MediaMentionUniqueAttributesBuilder.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    private Optional<Boolean> headline = Optional.empty();
+    private Optional<Boolean> authors = Optional.empty();
+    private Optional<Boolean> mediaOutletName = Optional.empty();
+    private Optional<Boolean> mediaType = Optional.empty();
+    private Optional<Boolean> scope = Optional.empty();
+    private Optional<Boolean> locale = Optional.empty();
+    private Optional<Boolean> language = Optional.empty();
+    private Optional<Boolean> coverageDate = Optional.empty();
+    private Optional<Boolean> datelineDate = Optional.empty();
+    private Optional<Boolean> category = Optional.empty();
```
[the remark this answers is not in the document]
Comment (willkoh) [reply Storage/src/main/java/storage/models/MediaMentionUniqueAttributesBuilder.java]: actually, not sure about doing it here. I think keeping them individually makes sense in the builder.

File: Storage/src/main/java/storage/controllers/SQLCommandController.java (hunk 3 above)
Comment (ameesh-shah) [reply Storage/src/main/java/storage/controllers/SQLCommandController.java]: +1 to will's comment

File: Storage/src/main/java/storage/models/MediaMentionUniqueAttributesBuilder.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (ameesh-shah) [reply Storage/src/main/java/storage/models/MediaMentionUniqueAttributesBuilder.java]: i think they should be kept individually

File: Storage/src/main/java/storage/models/SearchFilter.java [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+
+    public Optional<String[]> getKeywords() {
+        return keywords;
+    }
+
+    public void setKeywords(Optional<String[]> keywords) {
+        this.keywords = keywords;
+    }
+
+    public HashMap<String, Optional> getAllSingleValues() {
```
Comment (ameesh-shah) [inline Storage/src/main/java/storage/models/SearchFilter.java]: what's the reasoning behind maintaining the hashmap that you built as opposed to keeping the optional values separate?

File: Storage/src/main/java/storage/models/SearchResponse.java [hunk 13]
```diff
@@ -0,0 +1,22 @@
+package storage.models;
+
+import org.json.JSONArray;
+
+/**
+ * Created by spencerchang on 10/10/18.
+ */
+public class SearchResponse {
```
Comment (ameesh-shah) [inline Storage/src/main/java/storage/models/SearchResponse.java]: Wait aren't you just using a normal Response object in the API endpoint now tho

File: Storage/src/main/java/storage/processing/SearchHandler.java [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+                selectStatement.append(" WHERE ");
+            } else {
+                selectStatement.append(" AND ");
+            }
+            selectStatement.append(authorFilter);
+        }
+
+        if (body.getFilter().getKeywords().isPresent()) {
+            String keywordsFilter = Arrays.stream(body.getFilter().getKeywords().get())
+                .map((keyword) -> String.format("description LIKE '%%s%'", keyword))
```
Comment (ameesh-shah) [inline Storage/src/main/java/storage/processing/SearchHandler.java]: I believe this will have the apostrophes error that we encountered in batchProcessing. Can you make these single quotes escaped double quotes?

File: Storage/src/main/java/storage/controllers/DatabaseCRUDController.java [hunk 15]
```diff
@@ -77,4 +81,26 @@ public Response delete(@RequestParam(value="id") long id) {
...
+            // TODO log this, once we have logging
+            String errorMessage = e.getMessage();
+            return new Response(false, errorMessage);
+        }
+
+    }
+
+//    private String camelToSnakeCase(String inp) {
+//        return inp.replaceAll("(.)(\\p{Upper})", "$1_$2").toLowerCase();
+//    }
```
Comment (lucyrfox) [inline Storage/src/main/java/storage/controllers/DatabaseCRUDController.java]: Remove commented out code

(I don't think it should be needed, but if you want to keep this method for possible future use, throw it in util)

File: Storage/src/main/java/storage/models/SearchFilter.java [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+//        this.scope = scope;
+//        this.locale = locale;
+//        this.language = language;
+//        this.category = category;
+//        this.testDate = testDate;
+//        this.coverageRange = coverageRange;
+//        this.datelineRange = datelineRange;
+//        this.keywords = keywords;
+//        this.allSingleValues = allSingleValues;
+//    }
```
Comment (lucyrfox) [inline Storage/src/main/java/storage/models/SearchFilter.java]: remove commented out code

File: Storage/src/main/java/storage/models/SearchResponse.java (hunk 13 above)
Comment (lucyrfox) [reply Storage/src/main/java/storage/models/SearchResponse.java]: agreed — looks like you are using Response above.

also: We can add `JSONArray responseContent` to the Response object if you want to format in JSONArray instead of .toString()ing the JSON to return it (that would be better anyway, I just had trouble figuring it out)

File: Storage/src/main/java/storage/processing/SearchHandler.java (hunk 14 above)
Comment (lucyrfox) [reply Storage/src/main/java/storage/processing/SearchHandler.java]: I think it would be good to write some test cases just for this getSQLStatementForBody method. It's pretty complex and I'm sure has a lot of edge cases so would be useful to some isolated testing for it

Comment (lucyrfox) [reply Storage/src/main/java/storage/processing/SearchHandler.java]: oh looks like you set up the framework for that in `SearchHandlerTest` — would be good to do it more robustly tho

File: Storage/src/main/java/storage/models/MediaMentionUniqueAttributes.java (hunk 6 above)
Comment (spencerc99) [reply Storage/src/main/java/storage/models/MediaMentionUniqueAttributes.java]: like not have them as members? then they wont be recognized as needed from the json

File: Storage/src/main/java/storage/models/SearchFilter.java [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+
+    @JsonFormat(pattern="yyyy-MM-dd'T'HH:mm:ss.SSS")
+    private Optional<Timestamp[]> coverageRange;
+    @JsonFormat(pattern="yyyy-MM-dd'T'HH:mm:ss.SSS")
+    private Optional<Timestamp[]> datelineRange;
+    private Optional<String[]> keywords;
+    @JsonIgnore
+    private HashMap<String, Optional> allSingleValues;
```
[the remark this answers is not in the document]
Comment (spencerc99) [reply Storage/src/main/java/storage/models/SearchFilter.java]: wat it is in searchhandler

File: Storage/src/main/java/storage/processing/SearchHandler.java [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * Created by spencerchang on 10/5/18.
+ */
+public class SearchHandler {
+    public static String getSQLStatementForBody(SearchPayload body) {
+        StringBuilder selectStatement = new StringBuilder("SELECT ");
+        String selection = body.getAttributesToSelect().getAllValues().entrySet().stream()
+            .filter((e) -> e.getValue().isPresent())
+//            .map((e) -> String.format("%s%s", ((Boolean) e.getValue().get()) ? "UNIQUE " : "", e.getKey()))
```
[the remark this answers is not in the document]
Comment (spencerc99) [reply Storage/src/main/java/storage/processing/SearchHandler.java]: leaving because it is a todo remindr to handle the unique thing i mentioned in slack

File: Storage/src/main/java/storage/models/SearchFilter.java (hunk 12 above)
Comment (spencerc99) [reply Storage/src/main/java/storage/models/SearchFilter.java]: it is so i can generalize the building of the select and where statement instead of having an if case for everything

File: Storage/src/main/java/storage/models/SearchResponse.java (hunk 13 above)
Comment (spencerc99) [reply Storage/src/main/java/storage/models/SearchResponse.java]: yeah i am mentioned this in slack but was originally gonna use this but wanted to keep it consistent for now while we are all using response

Comment (spencerc99) [reply Storage/src/main/java/storage/models/SearchResponse.java]: we can migrate later for all the endpoints for consistency

File: Storage/src/main/java/storage/processing/SearchHandler.java (hunk 14 above)
Comment (spencerc99) [reply Storage/src/main/java/storage/processing/SearchHandler.java]: @ameesh-shah when is the case that you encountered this error? My example query that I tested is as follows which includes single quotes and works:

```
SELECT mediaOutletName, mediaType, headline FROM MediaMentions WHERE mediaOutletName = 'New York Times' AND coverageDate BETWEEN '2017-12-16 06:06:50.291' and '2018-12-16 06:06:50.291'
```

Comment (ameesh-shah) [reply Storage/src/main/java/storage/processing/SearchHandler.java]: e.g if you rename New York Times to New York Time's

Comment (ameesh-shah) [reply Storage/src/main/java/storage/processing/SearchHandler.java]: (when there are apostrophes in the actual fields themselves)

Comment (spencerc99) [reply Storage/src/main/java/storage/processing/SearchHandler.java]: @lucyrfox did a manual postman test for it and it works will write more extensive unit tests later (kinda need some official seeding for our development database with test data first for this though so w have standardized test data for unit tests)
